python/dcr_template.py

```diff
@@ -19,137 +20,156 @@
-                                                return_matrix=True)
-            if self.use_fft:
-                template = np.fft.ifft2(np.sum(self.model * dcr_kernel, axis=0))
-            else:
-                template = np.zeros((self.y_size, self.x_size))
-                weights = np.zeros((self.y_size, self.x_size))
-                pix_radius = self.kernel_size//2
-                for _j in range(self.y_size):
-                    for _i in range(self.x_size):
-                        # Deal with the edges later. Probably by padding the image with zeroes.
-                        if _i < pix_radius + 1:
-                            continue
-                        elif self.x_size - _i < pix_radius + 1:
-                            continue
-                        elif _j < pix_radius + 1:
-                            continue
-                        elif self.y_size - _j < pix_radius + 1:
-                            continue
-                        elif self.debug:
-                            if _i < x0: continue
-                            if _i > x0+dx: continue
-                            if _j < y0: continue
-                            if _j > y0+dy: continue
-                        model_vals = self._extract_model_vals(_j, _i, radius=pix_radius, fft=self.use_fft)
-                        template_vals = self._apply_dcr_kernel(dcr_kernel, model_vals)
-                        self._insert_template_vals(_j, _i, template_vals, template, weights,
-                                                   radius=pix_radius)
-                template[weights > 0] /= weights[weights > 0]
-                template[weights == 0] = 0.0
-                if add_noise:
-                    variance_level = np.median(exp.getMaskedImage().getVariance().getArray())
-                    rand_gen = np.random
-                    template += rand_gen.normal(scale=np.sqrt(variance_level), size=template.shape)
-
-            dataId_out = dataId[_img]
+                dcr_kernel = self._calc_offset_phase(**make_kernel_kwargs)
+
+            template = np.zeros((self.y_size, self.x_size))
+            weights = np.zeros((self.y_size, self.x_size))
+            pix_radius = self.kernel_size//2
+            for j in range(self.y_size):
+                for i in range(self.x_size):
+                    if self._edge_test(j, i):
+                        continue
+                    model_vals = self._extract_model_vals(j, i, radius=pix_radius,
+                                                          model=self.model, weights=self.weights)
+                    template_vals = self._apply_dcr_kernel(dcr_kernel, model_vals)
+                    self._insert_template_vals(j, i, template_vals, template=template, weights=weights,
+                                               radius=pix_radius, kernel=self.psf_avg)
+            template[weights > 0] /= weights[weights > 0]
```

Comment:
`weights` is updated for each exposure, and might be different. I've added some comments to clarify why `weights > 0` is checked each time.